core/chains/evm/gas/models.go

```diff
@@ -337,6 +344,56 @@ func (e *evmFeeEstimator) BumpFee(ctx context.Context, originalFee EvmFee, feeLi
 	return
 }
 
+func (e *evmFeeEstimator) estimateFeeLimit(ctx context.Context, fee EvmFee, feeLimit uint64, calldata []byte, toAddress *common.Address) (estimatedFeeLimit uint64, err error) {
+	// Use provided fee limit by default is EstimateGasLimit is disabled
+	if !e.geCfg.EstimateGasLimit() {
+		return commonfee.ApplyMultiplier(feeLimit, e.geCfg.LimitMultiplier())
+	}
+
+	// Create call msg for gas limit estimation
+	// Skip setting Gas to avoid capping the results of the estimation
+	callMsg := ethereum.CallMsg{
+		To:   toAddress,
+		Data: calldata,
+	}
+	if e.EIP1559Enabled {
+		callMsg.GasFeeCap = fee.DynamicFeeCap.ToInt()
+		callMsg.GasTipCap = fee.DynamicTipCap.ToInt()
+	} else {
+		// Set call msg fields with legacy fee field for gas limit estimation, if needed
+		callMsg.GasPrice = fee.Legacy.ToInt()
+	}
+	e.lggr.Debugw("estimating gas limit", "callMsg", callMsg)
+	gasUsed, estimateErr := e.ethClient.EstimateGas(ctx, callMsg)
+	if estimateErr != nil {
+		// Do not return error if estimate gas failed, we can still use the provided limit instead since it is an upper limit
+		e.lggr.Errorw("failed to estimate gas limit. falling back to provided gas limit.", "callMsg", callMsg, "providedGasLimit", feeLimit, "error", estimateErr)
+		estimatedFeeLimit, err = commonfee.ApplyMultiplier(feeLimit, e.geCfg.LimitMultiplier())
+		return
+	}
+	// Return error if estimated gas without the buffer exceeds the provided gas limit
+	// Transaction is destined to run out of gas and fail
+	if gasUsed > feeLimit {
```

Comment:
I noticed we already have a `LimitDefault` and a `LimitMax` config. Let's chat internally how to properly utilize them in this case.